### PR TITLE
[6.0.1] Use read-write lock to protect concurrent pool accesses

### LINKED ISSUE #26612: SqliteConnectionFactory.cs is not thread-safe
Reads on the `Dictionary` are not locked with the write operations, which is not supported.
Suggestion is to use a `ConcurrentDictionary` and remove some of the locks, or use a `ReaderWriterLockSlim`.

https://github.com/dotnet/efcore/blob/main/src/Microsoft.Data.Sqlite.Core/SqliteConnectionFactory.cs#L53

c.f. Dictionary source that is can throw an exception if the collection was changed while reads occur.
https://source.dot.net/#System.Private.CoreLib/Dictionary.cs,398

### PR DESCRIPTION
**Pull request by roji**
As this is being considered for servicing, I went with the least invasive/risky option, and introduced read/write locking. Using a ConcurrentDictionary is definitely possible, but would require other changes - specifically in how pruning is implemented. We can clean this up and improve for 7.0. I couldn't repro this in some quick testing efforts (tricky concurrency issue), if we think it's really worth it I can spend more time on it.

Fixes #26612

**Description**

Locking is missing when looking up a pool group in a Dictionary in SQLite's pooling implementation.

**Customer impact**

If code happens to look up a pool group at the same time as a modification of the pool group dictionary is happening (e.g. opening a connection with a new connection string), an exception may be thrown (or in theory, possibly undefined behavior).

**How found**

Reported by @sebastienros 

**Regression**

No, SQLite connection pooling is new in 6.0.

**Testing**

Testing this is somewhat tricky, and would not be completely reliable as this is a concurrency condition relying on a race condition.

**Risk**

Low, the current (partial) locking scheme has been augmented with read-write locking, changes are very localized. Added a quirk mode.

### CODE REVIEW

**Comment by sebastienros** [comment]
Why maintain the previous `Monitor` based lock, vs the `ReaderWriterLockSlim` which also handles the read path.

**Comment by roji** [comment]
@sebastienros the Monitor-based lock is only used under an AppContext-switch ("quirk"), this is something we do systematically when we patch, this way users can opt into the old behavior if there's trouble. In the main branch we'll only have the new reader/writer lock.

File: src/Microsoft.Data.Sqlite.Core/SqliteConnectionFactory.cs [hunk 1]
```diff
@@ -52,28 +53,65 @@ public SqliteConnectionInternal GetConnection(SqliteConnection outerConnection)
 
         public SqliteConnectionPoolGroup GetPoolGroup(string connectionString)
         {
-            if (!_poolGroups.TryGetValue(connectionString, out var poolGroup)
-                || (poolGroup.IsDisabled
-                    && !poolGroup.IsNonPooled))
+            var newLockingBehavior = !AppContext.TryGetSwitch("Microsoft.Data.Sqlite.Issue26612", out var enabled) || !enabled;
```
Comment (sebastienros) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnectionFactory.cs:56]: Why not in the ctor like the other switch? Though I understand the other switch has to be in the ctor.

Comment (roji) [reply src/Microsoft.Data.Sqlite.Core/SqliteConnectionFactory.cs:56]: Sure, I'll do that - thanks.

**Review by sebastienros — APPROVED** [review]

**Comment by sebastienros** [comment]
> this is something we do systematically when we patch

But this one is targeting main. I see the other switch is on `release/6.0`

I might have fixed the conflict incorrectly based on my interpretation.

**Comment by roji** [comment]
My bad, rebased on 6.0. We'll remove the quirk when we merge to main.

**Comment by sebastienros** [comment]
NB, AFAIK today is the last day to merge in release/6.0 for next service release

**Comment by roji** [comment]
Yeah, I know @sebastienros - just waiting for team signoff. We'll get this through.

**Review by smitpatel — APPROVED** [review]

**Review by bricelam — APPROVED** [review]

**Comment by roji** [comment]
Ah shit, that was a miss-click...

**Comment by dougbu** [comment]
Can revert the change if necessary. But, probably enough to cancel any internal builds including this commit.

**Comment by dougbu** [comment]
&hellip; that is, until you hear back from Tactics

**Comment by roji** [comment]
@AndriySvyryd what do you think? Sorry about the trouble...

**Comment by AndriySvyryd** [comment]
Need to revert it until it's approved

**Comment by dougbu** [comment]
> Need to revert it until it's approved

Please _also_ cancel any internal builds in the meantime. We don't want changes here to flow w/o that revert.

**Comment by AndriySvyryd** [comment]
> Please also cancel any internal builds in the meantime. We don't want changes here to flow w/o that revert.

Will do, it hasn't started yet

**Comment by roji** [comment]
Submitted #26696 to revert.

**Comment by roji** [comment]
Resubmitted this as #26697 after revertnig.

**Comment by Saibamen** [comment]
Should be moved to https://github.com/dotnet/efcore/milestone/130?closed=1

**Pull request merged**